luarules/gadgets/unit_builder_priority.lua

```diff
@@ -185,101 +216,223 @@ local function UpdatePassiveBuilders(teamID, interval)
 	-- calculate how much expense each passive con would require, and how much total expense the non-passive cons require
 	local nonPassiveConsTotalExpenseEnergy = 0
 	local nonPassiveConsTotalExpenseMetal = 0
-	local passiveConsExpense = {}
-	if not passiveCons[teamID] then
-		passiveCons[teamID] = {}
-	end
+
+	local passiveConsTotalExpenseEnergy = 0
+	local passiveConsTotalExpenseMetal = 0
+
+	passiveCons[teamID] = passiveCons[teamID] or {} -- alloc a table if not already
+
+	local passiveConsTeam = passiveCons[teamID]
+	if tracy then tracy.ZoneBeginN("UpdateStart") end 
+	local numPassiveCons = 0
+
+	local passiveExpense = {} -- once again, similar to the costIDOverride we are using a single table with offsets to store two values per unitID
+	-- this is about 800 us
 	if canBuild[teamID] then
-		for builderID in pairs(canBuild[teamID]) do
+		local canBuildTeam = canBuild[teamID]
+		for builderID in pairs(canBuildTeam) do
 			local builtUnit = spGetUnitIsBuilding(builderID)
-			local targetCosts = builtUnit and costID[builtUnit] or nil
-			if builtUnit and targetCosts and realBuildSpeed[builderID] then	-- added check for realBuildSpeed[builderID] else line below could error (unsure why)
-				local rate = realBuildSpeed[builderID] / targetCosts.buildTime
-				for _,resName in pairs(resTable) do
-					local expense = targetCosts[resName]
-					if expense <= 1 then -- metal maker costs 1 metal, don't stall because of that
-						expense = 0
-					else
-						expense = expense * rate
-					end
-					if passiveCons[teamID][builderID] then
-						if not passiveConsExpense[builderID] then
-							passiveConsExpense[builderID] = {energy=0, metal=expense}	-- because metal is set as first key
-						else
-							passiveConsExpense[builderID][resName] = expense
-						end
-						if not buildTargets[builtUnit] then
-							buildTargetOwners[builderID] = builtUnit
-							buildTargets[builtUnit] = true
-						end
-					else
-						if resName == 'energy' then
-							nonPassiveConsTotalExpenseEnergy = nonPassiveConsTotalExpenseEnergy + expense
-						else
-							nonPassiveConsTotalExpenseMetal = nonPassiveConsTotalExpenseMetal + expense
-						end
-					end
+			local expenseMetal = builtUnit and costIDOverride[builtUnit] or nil
+			local maxbuildspeed = maxBuildSpeed[builderID]
+			if builtUnit and expenseMetal and maxbuildspeed then	-- added check for maxBuildSpeed[builderID] else line below could error (unsure why), probably units that were newly created?
+
+				local expenseEnergy = costIDOverride[builtUnit + energyOffset]
+				local rate = maxbuildspeed / costIDOverride[builtUnit + buildTimeOffset]
+
+				-- TODO: redo solar and basic MM no-stall logic
+				
+				if expenseMetal <= 1 then 
+					expenseMetal = 0   -- metal maker costs 1 metal, don't stall because of that
+				else
+					expenseMetal = expenseMetal * rate
+				end
```

Comment:
Could be simplified a bit to `expenseMetal = expenseMetal <= 1 and 0 or expenseMetal * rate`